sal/tests/client/functional/test_operations.py

```diff
@@ -0,0 +1,243 @@
+        br = patched_client.list(path)
+
+        brr.assert_called_with('{0}/data/node?revision={1}'.format(host,
+                                                                   revision),
+                               verify=verify_https_cert)
+        assert isinstance(br, BranchReport)
+        assert br.timestamp == timestamp
+
+
+@pytest.mark.parametrize('path, revision, verify_https_cert, summary',
+                         [('/node', 0, False, True),
+                          ('/node:head', 0, True, False),
+                          ('/node:10', 10, False, False)])
+def test_get(host, patched_client, timestamp, branch_response, path, revision,
+             verify_https_cert, summary):
+
+    """
+    GIVEN
+        A server with data nodes
+    WHEN
+        A client performs a get operation
+    THEN
+        The client returns either a Branch or a DataObject
+    """
+
+    patched_client.verify_https_cert = verify_https_cert
+    obj = 'summary' if summary else 'full'
+
+    with patch('sal.client.main.requests.get',
+               return_value=branch_response) as br:
+        b = patched_client.get(path, summary=summary)
+
+        br.assert_called_with('{0}/data/node?object={1}&revision={2}'.format(
+            host, obj, revision), verify=verify_https_cert)
+        assert isinstance(b, Branch)
+
+
+@pytest.mark.parametrize('path, verify_https_cert',
+                         [('/node', False),
+                          ('/node:head', True),
+                          ('/node:0', False)])
+def test_put(host, patched_client, no_content_response, branch, path,
+             verify_https_cert):
+
+    """
+    GIVEN
+        A server with data nodes
+    WHEN
+        The client performs a put operation
+    THEN
+        A POST request for the specified path with a serialised Branch object
+        is sent to the host
+    """
+
+    patched_client.verify_https_cert = verify_https_cert
+
+    with patch('sal.client.main.requests.post',
+               return_value=no_content_response) as r:
+        patched_client.put(path, branch)
+
+        r.assert_called_with('{0}/data/node'.format(host),
+                             json=serialise(branch),
+                             verify=verify_https_cert)
+
+
+@pytest.mark.xfail(reason='SALCLient incorrectly sets source path',
+                   raises=AssertionError)
+@pytest.mark.parametrize('target, source, source_revision, verify_https_cert',
+                         [('/node', '/another/node', 0, False),
+                          ('/node:head', '/another/node:head', 0, True),
+                          ('/node', '/another/node:15', 15, True)])
+def test_copy(host, patched_client, no_content_response, target, source,
+              source_revision, verify_https_cert):
+
+    """
+    GIVEN
+        A server with data nodes
+    WHEN
+        The client performs a copy operation from a source to a target
+    THEN
+        A POST request 
```

Comment:
**question:** What does the POST request do?

Reply:
Whoops, I think this one had been left incomplete.  I've updated it.